src/primitives/hrp.rs

```diff
@@ -0,0 +1,406 @@
+// SPDX-License-Identifier: MIT
+
+//! Provides an `Hrp` type that represents the human-readable part of a bech32 encoded string.
+//!
+//! > The human-readable part, which is intended to convey the type of data, or anything else that
+//! > is relevant to the reader. This part MUST contain 1 to 83 US-ASCII characters, with each
+//! > character having a value in the range [33-126]. HRP validity may be further restricted by
+//! > specific applications.
+//!
+//! ref: [BIP-173](https://github.com/bitcoin/bips/blob/master/bip-0173.mediawiki#user-content-Bech32)
+
+#[cfg(feature = "alloc")]
+use alloc::string::String;
+use core::cmp::Ordering;
+use core::convert::TryInto;
+use core::fmt::{self, Write};
+use core::iter::{self, FusedIterator};
+use core::slice;
+
+use crate::Case;
+
+/// Maximum length of the human-readable part, as defined by BIP-173.
+const MAX_HRP_LEN: usize = 83;
+
+/// The human-readable part (human readable prefix before the '1' separator).
+#[derive(Clone, Copy, Debug)]
+pub struct Hrp {
+    /// ASCII byte values, guaranteed not to be mixed-case.
+    buf: [u8; MAX_HRP_LEN],
+    /// Number of characters currently stored in this HRP.
+    size: usize,
+}
+
+impl Hrp {
+    /// Parses the human-readable part checking it is valid as defined by [BIP-173].
+    ///
+    /// This does _not_ check that the `hrp` is an in-use HRP within Bitcoin (eg, "bc"), rather it
+    /// checks that the HRP string is valid as per the specification in [BIP-173]:
+    ///
+    /// > The human-readable part, which is intended to convey the type of data, or anything else that
+    /// > is relevant to the reader. This part MUST contain 1 to 83 US-ASCII characters, with each
+    /// > character having a value in the range [33-126]. HRP validity may be further restricted by
+    /// > specific applications.
+    ///
+    /// [BIP-173]: <https://github.com/bitcoin/bips/blob/master/bip-0173.mediawiki>
+    pub fn parse(hrp: &str) -> Result<Self, Error> { Ok(Self::parse_and_case(hrp)?.0) }
+
+    /// Returns the case as well as the parsed `Hrp`.
+    pub(crate) fn parse_and_case(hrp: &str) -> Result<(Self, Case), Error> {
+        use Error::*;
+
+        let mut new = Hrp { buf: [0_u8; MAX_HRP_LEN], size: 0 };
+
+        if hrp.is_empty() {
+            return Err(Empty);
+        }
+        if hrp.len() > MAX_HRP_LEN {
+            return Err(TooLong(hrp.len()));
+        }
+
+        let mut has_lower: bool = false;
+        let mut has_upper: bool = false;
+        for (i, c) in hrp.chars().enumerate() {
+            if !c.is_ascii() {
+                return Err(NonAsciiChar(c));
+            }
+            let b: u8 = c.try_into().expect("cannot fail for ASCII values");
```

Comment:
In 179868360b76d5d9a8d806dd5721d51d84e84b48:

I think you should just write `let b = c as u8; // cast OK as we just checked that `c` is an ASCII value` which is shorter and eliminates a branch, a panic, and a static string, none of which I really trust the compiler to get rid of.

Reply:
Sure thing, thanks for the reasoning.